lib/filewatch/watch.rb

```diff
@@ -117,127 +117,147 @@ def inode(path, stat)
+                # if the closed file changed, move it to the watched state
+                # not to active state because we want to use MAX_OPEN_FILES throttling.
+                watched_file.watch
+              end
+            rescue Errno::ENOENT
+              # file has gone away or we can't read it anymore.
+              file_deletable << path
+              @logger.debug? && @logger.debug("each: closed: stat failed: #{path}: (#{$!}), deleting from @files")
+            rescue => e
+              @logger.debug? && @logger.debug("each: closed: stat failed: #{path}: (#{e.inspect})")
             end
-          rescue Errno::ENOENT
-            # file has gone away or we can't read it anymore.
-            file_deleteable << path
-            @logger.debug? && @logger.debug("each: closed: stat failed: #{path}: (#{$!}), deleting from @files")
-          rescue => e
-            @logger.debug? && @logger.debug("each: closed: stat failed: #{path}: (#{e.inspect})")
           end
-        end
-
-        # look at the ignored to see if its changed
-        watched_files.select {|wf| wf.ignored? }.each do |watched_file|
-          path = watched_file.path
-          begin
-            stat = watched_file.restat
-            if watched_file.size_changed? || watched_file.inode_changed?(inode(path,stat))
-              # if the ignored file changed, move it to the watched state
-              # not to active state because we want to use MAX_OPEN_FILES throttling.
-              # this file has not been yielded to the block yet
-              # but we must have the tail to start from the end, so when the file
-              # was first ignored we updated the bytes_read to the stat.size at that time.
-              # by adding this to the sincedb so that the subsequent modify
-              # event can detect the change
-              watched_file.watch
-              yield(:unignore, watched_file)
+          return if quit?
+
+          # look at the ignored to see if its changed
+          watched_files.select {|wf| wf.ignored? }.each do |watched_file|
+            path = watched_file.path
+            break if quit?
+            begin
+              stat = watched_file.restat
+              if watched_file.size_changed? || watched_file.inode_changed?(inode(path,stat))
+                # if the ignored file changed, move it to the watched state
+                # not to active state because we want to use MAX_OPEN_FILES throttling.
+                # this file has not been yielded to the block yet
+                # but we must have the tail to start from the end, so when the file
+                # was first ignored we updated the bytes_read to the stat.size at that time.
+                # by adding this to the sincedb so that the subsequent modify
+                # event can detect the change
+                watched_file.watch
+                yield(:unignore, watched_file)
+              end
+            rescue Errno::ENOENT
```

Comment:
This repeated pattern is a mess.  Yeah, the restat operation will raise `Errno::ENOENT` but all bets are off with the `yield(:unignore)` call. If, between restat and the unignore callback, the file is deleted it could raise `Errno::ENOENT`
But you are correct, the log error messages are wrong.